File: JavaScript/cp/SlidingWindowMaximum.js
/* Sliding Window Maximum is a famous problem where we are given a list of integers
 nums and an integer k, return the maximum values of each sublist of length k. */


/* Function to get the maximum element of each sliding window of size k */
const slidingWindowMaximum = (nums, k) => {

    /* Sliding window */
    let window = [];

    /* Generating the initial window */
    for(let i=0; i<k; i++){
        
        /* Pushing the index to the window if there are no elements */
        if(window.length == 0){
            window.push(i);
            continue;
        }

        /* Getting the last index of our sliding window */
        let index = window[window.length-1];

        /* Arranging the elements in the window in descending order by removing the smaller ones */
        while(nums[i] > nums[index] && window.length > 0){
            window.pop();
            if(window.length > 0)
            index = window[window.length-1];
        }

        /* Pushing the current element to the window */
        window.push(i);
    }

    /* Answer array */
    let array = [];

    /* Storing the first maximum element in the array */
    array[0] = nums[window[0]];

    /* Generating k-1 windows */
    for(let i=k; i<nums.length; i++){

        let index = window[0];

        /* Removing indices which are out of bound */
        while(index <= i-k && window.length > 0){
            window.shift();
            if(window.length > 0)
            index = window[0];
        }

        if(window.length > 0)
        index = window[window.length-1];

        /* Arranging the elements in the window in ascending order by removing the ones which are smaller */
        while(nums[i] > nums[index] && window.length > 0){
            window.pop();
            if(window.length > 0)
            index = window[window.length-1];
        }

        window.push(i);

        /* Pushing the new maximum to the array */
        array[i-k+1] = nums[window[0]];
    }

    /* Returning the answer */
    return array;
}

//Using the readline() function of JavaScript to take input from the user
console.log("Enter the elements of of the array along with the size of the window");

//list is input array
let list = readline()
    .split(" ")
    .map((x) => parseInt(x));

// First n elements represent the elements in the array
let n = list.length-1;

// Last element of the list represents the size of the window k
let k = list[n];

/* Using list slicing and passing the first n elements to the
 function as the first parameter and k as the second parameter */
console.log("Array of maximum element in every window of size ", k, " is" ,slidingWindowMaximum(list.slice(0, n), k));

/* 
Example 1:

Input:

3 2 1 2 3 2

Output:

Array of maximum element in every window of size 2 is [3, 2, 2, 3]

Time Complexity : O(N)
Space Complexity : O(K)
*/

File: Python/cp/SlidingWindowMaximum.py
"""
Author: Guru Charan

Problem statement:
Given a list of integers "nums" and an integer k, return the maximum values
of each sublist of length k.

Constraints:
1 ≤ n ≤ 100,000 where n is the length of "nums".
1 ≤ k ≤ 100,000.

"""

from collections import deque


def findMaxInSubarray(nums, n , k) :

    '''
    paramaters:
    nums: input array
    n: size of the array
    k: window size
    output:
    returns a list (ans) containing the maximum of each window of size k

    '''
    ans = []

    # indices are stored in 'window'
    # with the maximum element at the front
    window = deque()

    # first k elements
    for i in range(k):

        # pop all the elements that are smaller than the current
        # element which are present towards its left (lower indices)
        while (window and (nums[i] >= nums[window[-1]])):
            window.pop()
        # adding the current element
        window.append(i)

    # storing the current answer
    ans.append(nums[window[0]])

    # getting the answer for rest of he array

    for i in range(k, n):

        # checking the window bounds
        if window[0] <= i-k:
            # remove an element from front
            window.popleft()
        while (window and nums[window[-1]] <= nums[i]):
            window.pop()
        window.append(i)
        # storing answer for current window
        ans.append(nums[window[0]])
    # returning the answer
    return ans


# DRIVER CODE

if __name__ == "__main__":

    size = int(input('enter the size of the array : '))

    numsArr = [int(i) for i in input('enter the elements of the array : ').split()]

    windowLen = int(input('enter the size of the window : '))

    print('The maximum values of each subarray of length k : ')
    print(*findMaxInSubarray(numsArr, size, windowLen))

ProblemDetails = '''Time Complexity
O(n) is the overall time complexity of the algorithm
O(k) is the overall space complexity of the algorithm since there can  only be k
elements in the window

Test Cases:

TC-1
Input:
enter the size of the array : 6
enter the elements of the array : 10 5 2 7 8 7
enter the size of the window : 3

Output:
The maximum values of each subarray of length k :
10 7 8 8

TC-2
Input:
enter the size of the array : 9
enter the elements of the array : 1 2 3 4 5 4 3 2 1
enter the size of the window : 3

Output:
The maximum values of each subarray of length k :
3 4 5 5 5 4 3'''
